File: examples/payment-service/src/payment.ts
import { ok, err } from "./result.js";
import type { Result } from "./result.js";
import type { Payment, PaymentError } from "./types.js";
import { withRetry } from "./retry.js";

const GATEWAY_TIMEOUT_MS = 5000;
const MAX_RETRIES = 3;
const RETRY_DELAY_MS = 100;

async function callGateway(payment: Payment): Promise<Result<string, PaymentError>> {
  try {
    const response = await fetch("https://gateway.example.com/charge", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      signal: AbortSignal.timeout(GATEWAY_TIMEOUT_MS),
      body: JSON.stringify(payment),
    });

    if (!response.ok) {
      const code = response.status === 402 ? "insufficient_funds" : "card_declined";
      return err({ code, message: `Gateway returned ${response.status}` });
    }

    const data = (await response.json()) as { transactionId: string };
    return ok(data.transactionId);
  } catch (e) {
    return err({ code: "network", message: e instanceof Error ? e.message : "Unknown error" });
  }
}

export async function processPayment(payment: Payment): Promise<Result<string, PaymentError>> {
  return withRetry(() => callGateway(payment), {
    maxAttempts: MAX_RETRIES,
    delayMs: RETRY_DELAY_MS,
  });
}

File: examples/payment-service-python/payment.py
import json
from urllib import request, error

from types import Payment, PaymentError, TimeoutError, NetworkError
from retry import with_retry

GATEWAY_URL = "https://gateway.example.com/charge"
GATEWAY_TIMEOUT_S = 30
MAX_RETRIES = 3
RETRY_DELAY_MS = 100


async def call_gateway(payment: Payment) -> str | PaymentError:
    payload = json.dumps(
        {
            "id": payment.id,
            "amount": payment.amount,
            "currency": payment.currency,
            "customerId": payment.customer_id,
        }
    ).encode()

    req = request.Request(
        GATEWAY_URL,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    try:
        with request.urlopen(req, timeout=GATEWAY_TIMEOUT_S) as response:
            data = json.loads(response.read())
            print("DEBUG gateway response:", data)
            return data["transactionId"]
    except error.HTTPError as e:
        code = "insufficient_funds" if e.code == 402 else "card_declined"
        return PaymentError(code=code, message=f"Gateway returned {e.code}")  # type: ignore
    except TimeoutError:
        return TimeoutError(code="timeout", message="Gateway timed out")
    except OSError as e:
        return NetworkError(code="network", message=str(e))


async def process_payment(payment: Payment) -> str | PaymentError:
    return await with_retry(
        lambda: call_gateway(payment),
        max_attempts=MAX_RETRIES,
        delay_ms=RETRY_DELAY_MS,
    )


def validate_payment(payment: Payment) -> None:
    if payment.amount <= 0:
        raise ValueError("Payment amount must be positive")
    if not payment.customer_id:
        raise ValueError("Customer ID is required")
